packages/jellyfish-transaction/src/script/dftx/dftx_masternode.ts

```diff
@@ -1,30 +1,42 @@
+import { SmartBuffer } from 'smart-buffer'
 import { BufferComposer, ComposableBuffer } from '../../buffer/buffer_composer'
 
 // Disabling no-return-assign makes the code cleaner with the setter and getter */
 /* eslint-disable no-return-assign */
 
 /**
- * CreateMasterNode DeFi Transaction
+ * CreateMasternode DeFi Transaction
  */
-export interface CreateMasterNode {
-  type: number // ----------------------------------| 1 byte, if 0x01 p2pkh, otherwise = p2wpkh
-  collateralPubKeyHash: string // ------------------| VarUInt{20 bytes}
-  operatorPubKeyHash?: string // -------------------| VarUInt{20 bytes}, p2pkh only, optional, use collateral pkh as default
+export interface CreateMasternode {
+  operatorType: number // --------------------------| 1 byte, 0x01 = p2pkh, 0x04 = p2wpkh
+  operatorAuthAddress: string // -------------------| VarUInt{20 bytes}
+  timelock?: number // -----------------------------| 2 bytes, 0x0401 = 5 years, 0x0802 = 10 years
 }
 
 /**
- * Composable CreateMasterNode, C stands for Composable.
+ * Composable CreateMasternode, C stands for Composable.
  * Immutable by design, bi-directional fromBuffer, toBuffer deep composer.
  */
-export class CCreateMasterNode extends ComposableBuffer<CreateMasterNode> {
+export class CCreateMasternode extends ComposableBuffer<CreateMasternode> {
   static OP_CODE = 0x43 // 'C'
   static OP_NAME = 'OP_DEFI_TX_CREATE_MASTER_NODE'
 
-  composers (cmn: CreateMasterNode): BufferComposer[] {
+  composers (cmn: CreateMasternode): BufferComposer[] {
     return [
-      ComposableBuffer.uInt8(() => cmn.type, v => cmn.type = v),
-      ComposableBuffer.hex(20, () => cmn.collateralPubKeyHash, v => cmn.collateralPubKeyHash = v),
-      ComposableBuffer.optionalHex(20, () => cmn.operatorPubKeyHash, v => cmn.operatorPubKeyHash = v)
+      ComposableBuffer.uInt8(() => cmn.operatorType, v => cmn.operatorType = v),
+      ComposableBuffer.hex(20, () => cmn.operatorAuthAddress, v => cmn.operatorAuthAddress = v),
+      {
+        fromBuffer: (buffer: SmartBuffer): void => {
+          if (buffer.remaining() > 0) {
+            cmn.timelock = buffer.readUInt16LE()
+          }
+        },
+        toBuffer: (buffer: SmartBuffer): void => {
+          if (cmn.timelock !== undefined) {
+            buffer.writeUInt16LE(cmn.timelock)
+          }
+        }
```

Comment:
Suggesting this here. not sure if i'm being short sighted here. if then please mention.
```suggestion
        fromBuffer: (buffer: SmartBuffer): void => {
          if (buffer.remaining() > 0) {
            cmn.timelock = buffer.readUInt16LE()
          }
        },
        toBuffer: (buffer: SmartBuffer): void => {
          if (cmn.timelock == undefined) {
              cmn.timelock = 0x0000
           }
           buffer.writeUInt16LE(cmn.timelock)
        }
```
Few  points.

- for pre or post eunosPaya `defid` can work with either `0x1c` or `0x1a` lengths.
- so on the jellyfish side, we can always serialize with `0x1c`, i.e with timelock.
- When deserializing at jellyfish side, we can support both `0x1c` or `0x1a` lengths with the above implementation. 
- But note that if we deserialize a pre eunosPaya tx and then serialize the same, we will get `0x1c` version with timelock `0x0000`

Reply:
i think only the last point will cause issue but not critical

> But note that if we deserialize a pre eunosPaya tx and then serialize the same, we will get 0x1c version with timelock 0x0000

so jellyfish user might need to take note while they're composing preEnuosPaya mn txid
i mean they need to manually fix the txid to get the data, seems not so good :thinking: 
```ts
const hex = '6a1a4466547843047b9275d2116bd2cd96a5f5b598a1470b99fc5c35' // <-- wrong
const hex = '6a1c4466547843047b9275d2116bd2cd96a5f5b598a1470b99fc5c350000' // <-- correct
const stack = toOPCodes(
      SmartBuffer.fromBuffer(Buffer.from(hex, 'hex'))
)
const buffer = toBuffer(stack)
```

Reply:
> i think only the last point will cause issue but not critical
> 
> > But note that if we deserialize a pre eunosPaya tx and then serialize the same, we will get 0x1c version with timelock 0x0000
> 
> so jellyfish user might need to take note while they're composing preEnuosPaya mn txid
> i mean they need to manually fix the txid to get the data, seems not so user-friendly 🤔
> 
> ```ts
> const hex = '6a1a4466547843047b9275d2116bd2cd96a5f5b598a1470b99fc5c35' // <-- wrong
> const hex = '6a1c4466547843047b9275d2116bd2cd96a5f5b598a1470b99fc5c350000' // <-- correct
> const stack = toOPCodes(
>       SmartBuffer.fromBuffer(Buffer.from(hex, 'hex'))
> )
> const buffer = toBuffer(stack)
> ```

do you mean the `hex` not `txid`? `hex` means hex string of the tx. in that case the above would be fine for both pre/post eunosPaya deserialization.
since you pointed out, the `txid` will definitely be changed if we deserialize a pre eunosPaya tx and serialize the same again in jellyfish side. if we have such usage, it might be a problem 🤔

Reply:
yea.. example use case in whale indexing.. user is required to handle the preEunosPaya mn hex... see whether we want opt to this approach

Reply:
hmm. then the ugly solution would be to incorporate `height` in dftx crafting layer or something similar to https://github.com/DeFiCh/jellyfish/pull/534#issuecomment-898264787 @fuxingloh

Reply:
Situation: post eunos (which will be all the way in the future), timelock undefined
still mandatory to write 0x0000 right?

Reply:
@surangap  @ivan-zynesis 
this is the issue
https://github.com/DeFiCh/whale/blob/8a050fc06bfc6cded44761952ab99b15f58d724e/src/module.indexer/model/dftx.indexer.ts#L60

Reply:
If we do not include any height info in dftx crafting layer, there will be no solution for actually able to create masternode buffer correctly both pre and post Eunos for timelock undefined case. Practically, it no longer requires to do pre Eunos, but forcing default value with have issue with the "bi-directional" test.

Reply:
> Practically, it no longer requires to do pre Eunos, but forcing default value with have issue with the "bi-directional" test.

I don't think this is a good practice. The better approach should be able to handle pre an post cases.

Reply:
it's actually not that critical, it's just a "ux", no point to having long discussion but dangling the PR also blocking the project
either `force default` or `optional`
i follow majority

Reply:
Let's go with optional now

Although it's inconsistent, it's something bearable on the jellyfish side. To properly look into this, we need to explore this in Q4 with an indexer rebuild.

It's not a huge issue for crafting transactions because `timelock` can be enforced on `jellyfish-transaction-builder` making it non optional.

Reply:
My question was actually if post Eunos defid will reject timelock undefined (missing 0x0000 byte), as long as that do not happen, this looks good.